src/Haskell/Ide/Engine/LSP/CodeActions.hs

```diff
@@ -48,45 +48,42 @@ handleCodeActionReq tn req = do
         let reqs = map (\f -> f docId maybeRootDir range context) providers
         in collectRequests reqs (send . concat)
 
-  makeRequest (IReq tn (req ^. J.id) providersCb getProviders)
+  collectRequest getProviders providersCb
 
   where
     params = req ^. J.params
     docUri = params ^. J.textDocument . J.uri
     range = params ^. J.range
     context = params ^. J.context
 
-    wrapCodeAction :: J.CodeAction -> R (Maybe J.CommandOrCodeAction)
+    wrapCodeAction :: J.CodeAction -> R J.CommandOrCodeAction
     wrapCodeAction action' = do
       prefix <- asks commandPrefixer :: R (T.Text -> T.Text)
-
-      (C.ClientCapabilities _ textDocCaps _) <- asksLspFuncs Core.clientCapabilities
-      let literalSupport = textDocCaps >>= C._codeAction >>= C._codeActionLiteralSupport
-          action :: J.CodeAction
+ 
+      let action :: J.CodeAction
           action = (J.command . _Just . J.command %~ prefix) action'
+      supported <- asksLspFuncs $ has
+        $ to Core.clientCapabilities . to C._textDocument . _Just . to C._codeAction . _Just . to C._codeActionLiteralSupport . _Just
+        . maybe united (\k -> to C._codeActionKind . to C._valueSet . traverse . only k) (action ^. J.kind)
 
-      case literalSupport of
-        Nothing ->
-            let cmd = J.Command (action ^. J.title) cmdName (Just cmdParams)
-                cmdName = prefix "hie:fallbackCodeAction"
-                cmdParams = J.List [J.toJSON (FallbackCodeActionParams (action ^. J.edit) (action ^. J.command))]
-              in return $ Just (J.CommandOrCodeActionCommand cmd)
-        Just _ -> return $ Just (J.CommandOrCodeActionCodeAction action)
+      return $ if supported
+        then J.CommandOrCodeActionCodeAction action
+        else
+          let cmd = J.Command (action ^. J.title) cmdName (Just cmdParams)
+              cmdName = prefix "hie:fallbackCodeAction"
+              cmdParams = J.List [J.toJSON (FallbackCodeActionParams (action ^. J.edit) (action ^. J.command))]
+            in J.CommandOrCodeActionCommand cmd
 
     send :: [J.CodeAction] -> R ()
-    send codeActions = do
-      body <- J.List . catMaybes <$> mapM wrapCodeAction codeActions
-      reactorSend $ RspCodeAction $ Core.makeResponseMessage req body
+    send = reactorSend . RspCodeAction . Core.makeResponseMessage req . J.List <=< mapM wrapCodeAction
 
     -- | Execute multiple ide requests sequentially
     collectRequests :: [IdeM (IdeResponse a)] -- ^ The requests to make
                     -> ([a] -> R ())     -- ^ Callback with the request inputs and results
                     -> R ()
-    collectRequests = go []
-      where
-        go acc [] callback = callback acc
-        go acc (x:xs) callback =
-          let reqCallback result = go (acc ++ [result]) xs callback
-          in makeRequest $ IReq tn (req ^. J.id) reqCallback x
-
+    collectRequests = alaf ContT traverse collectRequest
```

Comment:
Wow this is cool. Very succinct!

Reply:
I am biased toward understandability rather than succinctness.

To me, it is not at all clear what the proposed code does, and it requires a side investigation to work it out.